Support LPC55S36 in Zephyr

NeilChen93 opened the pull request.

danieldegrasse committed:
drivers: pinctrl: update lpc iocon to support LPC55s3x

LPC55s3x family has an additional analog mode bit available. Update
IOCON binding and IOCON driver to support setting this bit via a pin
control property.

Signed-off-by: Daniel DeGrasse <daniel.degrasse@nxp.com>

NeilChen93 commented:
Hi all, this PR is prepared for LPC55S36, please help review, thanks a lot. And the file pinctrl_soc.h in location zephyr\soc\arm\nxp_lpc\lpc55xxx\pinctrl_soc.h need to be updated, because LPC55S36’s analog switch input control bit name is different with other LPC55xx chips. Daniel will be responsible to update it. Then LPC55S36's hello_world example can build and run successfully.

dleach02 commented:
@NeilChen93, we need to clean up the commit history here. The "merge" of main commit needs to go away where your branch would use a rebase from main to get your branch up to date.

dleach02 reviewed (changes requested):
Use rebase to update branch with main.

dleach02 commented on soc/arm/nxp_lpc/lpc55xxx/soc.c:
```diff
@@ -59,7 +59,8 @@ const pll_setup_t pll0Setup = {
-	defined(CONFIG_SOC_LPC55S28) || defined(CONFIG_SOC_LPC55S69_CPU0)
+	defined(CONFIG_SOC_LPC55S28) || defined(CONFIG_SOC_LPC55S36) || \
+	defined(CONFIG_SOC_LPC55S69_CPU0)
     /*!< Set up the clock sources */
```
Coding standard for Zephyr uses tabs instead of spaces

dleach02 replied:
This applies to the comment block as well

mmahadevan108 commented:
Please add the pinctrl related updates made by @danieldegrasse.

mmahadevan108 commented on boards/arm/lpcxpresso55s36/Kconfig.defconfig:
```diff
@@ -0,0 +1,11 @@
+# LPCXpresso55S36 board
+
+# Copyright (c) 2022 NXP
```
NXP guidance on Copyright is to drop the `(c)`

NeilChen93 replied:
Thanks for reminder.

mmahadevan108 commented on boards/arm/lpcxpresso55s36/lpcxpresso55s36_common.dtsi:
```diff
@@ -0,0 +1,114 @@
+/*
```
Do we need a `lpcxpresso55s36_common.dtsi` file. Seems like this can be move to `lpcxpresso55s36.dtsi`.

NeilChen93 replied:
I will try to move this file into lpcxpresso55s36.dtsi

danieldegrasse commented:
@NeilChen93 Do you want me to open a separate PR with the updates for the pin control driver, or integrate those changes here?

NeilChen93 commented:
Hi @danieldegrasse, it is both OK for me. I prefer you to open a separate PR.

danieldegrasse commented:
> Hi @danieldegrasse, it is both OK for me. I prefer you to open a separate PR.
@NeilChen93 #48193 should contain the required updates for the LPC55s3x family pin control driver

NeilChen93 commented:
Hi all, I cherry pick daniel's commit into this PR. And LPC55S36 examples should can build and run now, please help review and check again. Thanks a lot.

dleach02 reviewed (changes requested):
@NeilChen93, the image for the board is a LPC55S16 board and not the LPC5536 board

dleach02 commented on soc/arm/nxp_lpc/lpc55xxx/Kconfig.soc:
```diff
@@ -29,6 +30,7 @@ config SOC_LPC55S16
 	select CLOCK_CONTROL
 	select HAS_MCUX_IAP if !TRUSTED_EXECUTION_NONSECURE
 	select HAS_MCUX_MCAN
+	select HAS_MCUX_RNG
```
We should focus this PR only on the LPC55S36. If we want to add the select of RNG to other platforms lets do that in a standalone PR that also connects the entropy driver.

NeilChen93 replied:
Hi @dleach02 , LPC55S36 don't support RNG, so I removed LPC55XXX common config HAS_MCUX_RNG in file "soc/arm/nxp_lpc/lpc55xxx/Kconfig.series", and add the select of RNG separately for other platforms. Is there any other better way to remove RNG support for LPC55S36? Thanks a lot.

dleach02 replied:
okay. That makes sense.

dleach02 commented:
@NeilChen93, please address the CI failures

galak commented on dts/arm/nxp/nxp_lpc55S3x_common.dtsi:
```diff
@@ -0,0 +1,245 @@
+	syscon: syscon@0 {
+		compatible = "nxp,lpc-syscon";
+		reg = <0x0 0x1000>;
+		label = "SYSCON";
```
remove `label`

NeilChen93 replied:
Already removed it.

galak commented on dts/arm/nxp/nxp_lpc55S3x_common.dtsi:
```diff
@@ -0,0 +1,245 @@
+
+	iap: flash-controller@34000 {
+		compatible = "nxp,lpc-iap";
+		label = "FLASH_IAP";
```
remove `label` property

NeilChen93 replied:
Already removed it.

galak commented on dts/arm/nxp/nxp_lpc55S3x_common.dtsi:
```diff
@@ -0,0 +1,245 @@
+
+		flash0: flash@0 {
+			compatible = "soc-nv-flash";
+			label = "MCUX_FLASH";
```
remove `label` property

NeilChen93 replied:
Already removed it.

galak commented on dts/arm/nxp/nxp_lpc55S3x_common.dtsi:
```diff
@@ -0,0 +1,245 @@
+		compatible = "nxp,lpc-gpio";
+		reg = <0x8c000 0x278c>;
+		interrupts = <4 2>,<5 2>,<6 2>,<7 2>;
+		label = "GPIO_0";
```
remove `label` property

NeilChen93 replied:
Already removed it.

galak commented on dts/arm/nxp/nxp_lpc55S3x_common.dtsi:
```diff
@@ -0,0 +1,245 @@
+		compatible = "nxp,lpc-gpio";
+		reg = <0x8c000 0x278c>;
+		interrupts = <32 2>,<33 2>,<34 2>,<35 2>;
+		label = "GPIO_1";
```
remove `label` property

NeilChen93 replied:
Already removed it.

galak commented on dts/arm/nxp/nxp_lpc55S3x_common.dtsi:
```diff
@@ -0,0 +1,245 @@
+		compatible = "nxp,lpc-gpio";
+		reg = <0x8c000 0x278c>;
+		#interrupts = <32 2>,<33 2>,<34 2>,<35 2>;
+		label = "GPIO_2";
```
remove `label` property

NeilChen93 replied:
Already removed it.

galak commented on dts/arm/nxp/nxp_lpc55S3x_common.dtsi:
```diff
@@ -0,0 +1,245 @@
+		reg = <0x86000 0x1000>;
+		interrupts = <14 0>;
+		clocks = <&syscon MCUX_FLEXCOMM0_CLK>;
+		label = "FLEXCOMM_0";
```
remove `label` property

NeilChen93 replied:
Already removed it.

galak commented on dts/arm/nxp/nxp_lpc55S3x_common.dtsi:
```diff
@@ -0,0 +1,245 @@
+/*
```
remove `label` property from all nodes in here

NeilChen93 replied:
Already removed it.

galak commented on soc/arm/nxp_lpc/lpc55xxx/Kconfig.defconfig.lpc55S36:
```diff
@@ -0,0 +1,27 @@
+
+config SOC
+	default "lpc55S36"
+
```
remove PINMUX_MCUX_LPC, GPIO_MCUX_LPC, and UART_MCUX_FLEXCOMM.  They aren't needed.

NeilChen93 replied:
Already removed it.

galak reviewed (changes requested):
* Remove `label` usage from soc dtsi
* Some Kconfig updates

yvanderv commented on boards/arm/lpcxpresso55s36/lpcxpresso55s36-pinctrl.dtsi:
```diff
@@ -0,0 +1,51 @@
+ * NOTE: File generated by lpc_cfg_utils.py
+ * from LPC55S36.mex
+ *
+ * Copyright 2022, NXP
```
Why a comma here?

NeilChen93 replied:
Updated it.

yvanderv commented on dts/arm/nxp/nxp_lpc55S3x_common.dtsi:
```diff
@@ -0,0 +1,245 @@
+		reg = <0x20000000 DT_SIZE_K(32)>;
+	};
+
+	sram2: memory@20008000 {
```
Why not SRAM1?

dleach02 replied:
@NeilChen93, lets do something similar that was done for the other LPC parts (with related changes to the ns dtsi):

	/* lpc55_3x Memory configurations:
	 *
	 * LPC55x36: RAMX: 16K, SRAM0: 16K, SRAM1: 16K, SRAM2: 32K, SRAM3: 32K, SRAM4: 16K
	 */
	sramx: memory@4000000 {
		compatible = "mmio-sram";
		reg = <0x04000000 DT_SIZE_K(16)>;
	};
	sram0: memory@20000000 {
		compatible = "mmio-sram";
		reg = <0x20000000 DT_SIZE_K(16)>;
	};
	sram1: memory@20004000 {
		compatible = "mmio-sram";
		reg = <0x20004000 DT_SIZE_K(16)>;
	};
	sram2: memory@20008000 {
		compatible = "mmio-sram";
		reg = <0x20008000 DT_SIZE_K(32)>;
	};
	sram3: memory@20010000 {
		compatible = "mmio-sram";
		reg = <0x20010000 DT_SIZE_K(32)>;
	};
	sram4: memory@20018000 {
		compatible = "mmio-sram";
		reg = <0x20018000 DT_SIZE_K(16)>;
	};

NeilChen93 replied:
Thanks dleach.

dleach02 commented on dts/arm/nxp/nxp_lpc55S36_ns.dtsi:
```diff
@@ -0,0 +1,19 @@
+	};
+};
+
+#include "nxp_lpc55S3x_common.dtsi"
```
After this line, add the following:

/*
 * lpc55S_36:
 * Combine SRAM0, SRAM1, SRAM2, SRAM3, SRAM4 for total of 112K RAM
 */
&sram0 {
	compatible = "mmio-sram";
	reg = <0x20000000 DT_SIZE_K(112)>;
};

NeilChen93 replied:
Got it, thanks for reminder.

NeilChen93 committed:
soc: arm: nxp_lpc: lpc55xxx: LPC55S36: add device support

- Add soc information for LPC55S36

Signed-off-by: Neil Chen <cheng.chen_1@nxp.com>

NeilChen93 commented on boards/arm/lpcxpresso55s36/doc/index.rst:
```diff
@@ -0,0 +1,201 @@
+Supported Features
+==================
+
+The lpcxpresso55s36 board configuration supports the following
```
Updated it.

dleach02 commented on dts/arm/nxp/nxp_lpc55S36_ns.dtsi:
```diff
@@ -0,0 +1,28 @@
+&sram0 {
+	compatible = "mmio-sram";
+	reg = <0x20000000 DT_SIZE_K(112)>;
+};
```
need newline after

dleach02 replied:
@NeilChen93, you need a newline at the end of the file. This will clean up one of the conformance test errors in CI

yvanderv committed:
soc: arm: nxp_lpc: lpc55xxx: disabling LPC55S36 IAP support

This is temporary until support has been enabled. lpc55s36's IAP controller
is different and does not match previous implementations.
The Zephyr driver uses a data structure matching the IAP controller to
implement a feature that should normally come out of the ROM.

Since IAP support is not in yet, disabling the option in the SoC's Kconfig
and disabling the node in the device tree is the proper way to go forward.

Signed-off-by: Yves Vandervennet <yves.vandervennet@nxp.com>

NeilChen93 committed:
dts: arm: nxp: Add dtsi for LPC55S36

The default RAM configuration for LPC55S36:
- sramx 0x4000000  16K
- sram0 0x20000000 16K
- sram1 0x20004000 16K
- sram2 0x20008000 32K
- sram3 0x20010000 32K
- sram4 0x20018000 16K

Signed-off-by: Neil Chen <cheng.chen_1@nxp.com>

NeilChen93 committed:
board: arm: Add board support for lpcxpresso55s36

- Add pinctrl, dts, doc

Signed-off-by: Neil Chen <cheng.chen_1@nxp.com>

NeilChen93 committed:
west.yml: Update NXP HAL to get the latest SDK

 - Update the NXP HAL SDK

Signed-off-by: Neil Chen <cheng.chen_1@nxp.com>

NeilChen93 commented:
Hi @dleach02 , @galak , I already updated according to your comments and this PR has passed all CI test, so please help check again and approve.

yvanderv reviewed (commented):
Looks good to me
